File: php/112-path-sum.php
<?php
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     public $val = null;
 *     public $left = null;
 *     public $right = null;
 *     function __construct($value) { $this->val = $value; }
 * }
 */
class Solution 
{
    /**
     * @param TreeNode $root
     * @param Integer $sum
     * @return Boolean
     */
    public function hasPathSum($root, $sum) 
    {
        if ($root == null) {
            return false;
        }
        if ($root->left == null && $root->right == null) {
            return $root->val == $sum;
        }
        $need = $sum - $root->val;
        if ($this->hasPathSum($root->left, $need)) {
            return true;
        }
        if ($this->hasPathSum($root->right, $need)) {
            return true;
        }
        return false;
    }
}

File: go/112-path-sum.go
package main

/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func hasPathSum(root *TreeNode, sum int) bool {
	if root == nil {
		return false
	}
	if root.Left == nil && root.Right == nil {
		if root.Val == sum {
			return true
		}
		return false
	}
	need := sum - root.Val
	if hasPathSum(root.Left, need) {
		return true
	}
	if hasPathSum(root.Right, need) {
		return true
	}
	return false
}
